File: flutter/lib/services/rebrickable_api_client_impl.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import '../models/api_response.dart';
import '../models/lego_models.dart';
import 'rebrickable_api_client.dart';
import 'rebrickable_exceptions.dart';

/// Implementation of RebrickableApiClient using HTTP package
class RebrickableApiClientImpl implements RebrickableApiClient {
  static const String _baseUrl = 'https://rebrickable.com/api/v3/lego/';
  static const int _maxPageSize = 1000;

  final String _apiKey;
  final http.Client _httpClient;

  RebrickableApiClientImpl({
    required String apiKey,
    http.Client? httpClient,
  })  : _apiKey = apiKey,
        _httpClient = httpClient ?? http.Client();

  @override
  Future<ApiResponse<LegoSet>> getSets({
    String? search,
    int page = 1,
    int pageSize = 20,
    int? themeId,
    int? minYear,
    int? maxYear,
    int? minParts,
    int? maxParts,
  }) async {
    final params = <String, String>{
      'key': _apiKey,
      'page': page.toString(),
      'page_size': pageSize.clamp(1, _maxPageSize).toString(),
    };

    if (search != null) params['search'] = search;
    if (themeId != null) params['theme_id'] = themeId.toString();
    if (minYear != null) params['min_year'] = minYear.toString();
    if (maxYear != null) params['max_year'] = maxYear.toString();
    if (minParts != null) params['min_parts'] = minParts.toString();
    if (maxParts != null) params['max_parts'] = maxParts.toString();

    final uri = Uri.parse('${_baseUrl}sets/').replace(queryParameters: params);
    final response = await _makeRequest(uri);
    
    return ApiResponse.fromJson(
      response,
      (json) => LegoSet.fromJson(json as Map<String, dynamic>),
    );
  }

  @override
  Future<LegoSet> getSet(String setNum) async {
    final uri = Uri.parse('${_baseUrl}sets/$setNum/').replace(
      queryParameters: {'key': _apiKey},
    );
    final response = await _makeRequest(uri);
    return LegoSet.fromJson(response);
  }

  @override
  Future<ApiResponse<LegoPart>> getParts({
    String? search,
    int page = 1,
    int pageSize = 20,
    int? catId,
  }) async {
    final params = <String, String>{
      'key': _apiKey,
      'page': page.toString(),
      'page_size': pageSize.clamp(1, _maxPageSize).toString(),
    };

    if (search != null) params['search'] = search;
    if (catId != null) params['part_cat_id'] = catId.toString();

    final uri = Uri.parse('${_baseUrl}parts/').replace(queryParameters: params);
    final response = await _makeRequest(uri);
    
    return ApiResponse.fromJson(
      response,
      (json) => LegoPart.fromJson(json as Map<String, dynamic>),
    );
  }

  @override
  Future<LegoPart> getPart(String partNum) async {
    final uri = Uri.parse('${_baseUrl}parts/$partNum/').replace(
      queryParameters: {'key': _apiKey},
    );
    final response = await _makeRequest(uri);
    return LegoPart.fromJson(response);
  }

  @override
  Future<ApiResponse<LegoTheme>> getThemes({
    int page = 1,
    int pageSize = 20,
  }) async {
    final params = <String, String>{
      'key': _apiKey,
      'page': page.toString(),
      'page_size': pageSize.clamp(1, _maxPageSize).toString(),
    };

    final uri = Uri.parse('${_baseUrl}themes/').replace(queryParameters: params);
    final response = await _makeRequest(uri);
    
    return ApiResponse.fromJson(
      response,
      (json) => LegoTheme.fromJson(json as Map<String, dynamic>),
    );
  }

  @override
  Future<ApiResponse<LegoColor>> getColors({
    int page = 1,
    int pageSize = 20,
  }) async {
    final params = <String, String>{
      'key': _apiKey,
      'page': page.toString(),
      'page_size': pageSize.clamp(1, _maxPageSize).toString(),
    };

    final uri = Uri.parse('${_baseUrl}colors/').replace(queryParameters: params);
    final response = await _makeRequest(uri);
    
    return ApiResponse.fromJson(
      response,
      (json) => LegoColor.fromJson(json as Map<String, dynamic>),
    );
  }

  /// Make HTTP request and handle common errors
  Future<Map<String, dynamic>> _makeRequest(Uri uri) async {
    try {
      final response = await _httpClient.get(uri);
      return _handleResponse(response);
    } on SocketException catch (e) {
      throw NetworkException('Network error: ${e.message}', e);
    } on HttpException catch (e) {
      throw NetworkException('HTTP error: ${e.message}', e);
    } catch (e) {
      throw NetworkException('Unknown network error: ${e.toString()}', Exception(e.toString()));
    }
  }

  /// Handle HTTP response and parse JSON
  Map<String, dynamic> _handleResponse(http.Response response) {
    final statusCode = response.statusCode;
    
    try {
      final jsonData = json.decode(response.body) as Map<String, dynamic>;
      
      if (statusCode >= 200 && statusCode < 300) {
        return jsonData;
      }
      
      // Handle error responses
      final error = ApiError.fromJson(jsonData);
      final message = error.detail ?? 'API error occurred';
      
      switch (statusCode) {
        case 401:
          throw AuthenticationException('Invalid API key: $message');
        case 403:
          throw AuthenticationException('Access forbidden: $message');
        case 404:
          throw NotFoundException('Resource not found: $message');
        case 429:
          throw RateLimitException('Rate limit exceeded: $message');
        case >= 500:
          throw ServerException('Server error: $message');
        default:
          throw ApiException('API error: $message', statusCode);
      }
    } on FormatException catch (e) {
      throw ApiException('Invalid JSON response: ${e.message}', statusCode, e);
    }
  }

  @override
  void dispose() {
    _httpClient.close();
  }
}

File: shared/src/commonMain/kotlin/eu/mpwg/brixie/shared/api/RebrickableApiClientImpl.kt
package eu.mpwg.brixie.shared.api

import eu.mpwg.brixie.shared.api.models.ApiError
import eu.mpwg.brixie.shared.api.models.ApiResponse
import eu.mpwg.brixie.shared.api.models.LegoColor
import eu.mpwg.brixie.shared.api.models.LegoPart
import eu.mpwg.brixie.shared.api.models.LegoSet
import eu.mpwg.brixie.shared.api.models.LegoTheme
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.defaultRequest
import io.ktor.client.plugins.logging.LogLevel
import io.ktor.client.plugins.logging.Logger
import io.ktor.client.plugins.logging.Logging
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.client.statement.HttpResponse
import io.ktor.http.HttpStatusCode
import io.ktor.http.isSuccess
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

/**
 * Implementation of RebrickableApiClient using Ktor HTTP client
 */
class RebrickableApiClientImpl(
    private val apiKey: String = getApiKey()
) : RebrickableApiClient {

    companion object {
        private const val TAG = "RebrickableApiClient"
        private const val BASE_URL = "https://rebrickable.com/api/v3/lego/"
        private const val MAX_PAGE_SIZE = 1000
    }

    private val httpClient = HttpClient(createHttpClientEngine()) {
        defaultRequest {
            url(BASE_URL)
        }
        
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient = true
                encodeDefaults = false
            })
        }
        
        install(Logging) {
            logger = object : Logger {
                override fun log(message: String) {
                    platformLog(TAG, message)
                }
            }
            level = if (isDebugMode()) LogLevel.INFO else LogLevel.NONE
        }
    }

    override suspend fun getSets(
        search: String?,
        page: Int,
        pageSize: Int,
        themeId: Int?,
        minYear: Int?,
        maxYear: Int?,
        minParts: Int?,
        maxParts: Int?
    ): Result<ApiResponse<LegoSet>> = withContext(Dispatchers.Default) {
        try {
            val response: HttpResponse = httpClient.get("sets/") {
                parameter("key", apiKey)
                parameter("page", page)
                parameter("page_size", pageSize.coerceAtMost(MAX_PAGE_SIZE))
                search?.let { parameter("search", it) }
                themeId?.let { parameter("theme_id", it) }
                minYear?.let { parameter("min_year", it) }
                maxYear?.let { parameter("max_year", it) }
                minParts?.let { parameter("min_parts", it) }
                maxParts?.let { parameter("max_parts", it) }
            }
            handleResponse(response)
        } catch (e: Exception) {
            Result.failure(handleException(e))
        }
    }

    override suspend fun getSet(setNum: String): Result<LegoSet> = withContext(Dispatchers.Default) {
        try {
            val response: HttpResponse = httpClient.get("sets/$setNum/") {
                parameter("key", apiKey)
            }
            handleResponse(response)
        } catch (e: Exception) {
            Result.failure(handleException(e))
        }
    }

    override suspend fun getParts(
        search: String?,
        page: Int,
        pageSize: Int,
        catId: Int?
    ): Result<ApiResponse<LegoPart>> = withContext(Dispatchers.Default) {
        try {
            val response: HttpResponse = httpClient.get("parts/") {
                parameter("key", apiKey)
                parameter("page", page)
                parameter("page_size", pageSize.coerceAtMost(MAX_PAGE_SIZE))
                search?.let { parameter("search", it) }
                catId?.let { parameter("part_cat_id", it) }
            }
            handleResponse(response)
        } catch (e: Exception) {
            Result.failure(handleException(e))
        }
    }

    override suspend fun getPart(partNum: String): Result<LegoPart> = withContext(Dispatchers.Default) {
        try {
            val response: HttpResponse = httpClient.get("parts/$partNum/") {
                parameter("key", apiKey)
            }
            handleResponse(response)
        } catch (e: Exception) {
            Result.failure(handleException(e))
        }
    }

    override suspend fun getThemes(
        page: Int,
        pageSize: Int
    ): Result<ApiResponse<LegoTheme>> = withContext(Dispatchers.Default) {
        try {
            val response: HttpResponse = httpClient.get("themes/") {
                parameter("key", apiKey)
                parameter("page", page)
                parameter("page_size", pageSize.coerceAtMost(MAX_PAGE_SIZE))
            }
            handleResponse(response)
        } catch (e: Exception) {
            Result.failure(handleException(e))
        }
    }

    override suspend fun getColors(
        page: Int,
        pageSize: Int
    ): Result<ApiResponse<LegoColor>> = withContext(Dispatchers.Default) {
        try {
            val response: HttpResponse = httpClient.get("colors/") {
                parameter("key", apiKey)
                parameter("page", page)
                parameter("page_size", pageSize.coerceAtMost(MAX_PAGE_SIZE))
            }
            handleResponse(response)
        } catch (e: Exception) {
            Result.failure(handleException(e))
        }
    }

    /**
     * Generic response handler for all API calls
     */
    private suspend inline fun <reified T> handleResponse(response: HttpResponse): Result<T> {
        return try {
            if (response.status.isSuccess()) {
                val data: T = response.body()
                Result.success(data)
            } else {
                val error = when (response.status) {
                    HttpStatusCode.Unauthorized -> AuthenticationException("Invalid API key")
                    HttpStatusCode.NotFound -> NotFoundException("Resource not found")
                    HttpStatusCode.BadRequest -> BadRequestException("Invalid request parameters")
                    HttpStatusCode.TooManyRequests -> RateLimitException("Rate limit exceeded")
                    else -> {
                        try {
                            val errorBody: ApiError = response.body()
                            ServerException(errorBody.detail ?: "Server error: ${response.status}")
                        } catch (e: Exception) {
                            ServerException("Server error: ${response.status}")
                        }
                    }
                }
                Result.failure(error)
            }
        } catch (e: Exception) {
            Result.failure(ParseException("Failed to parse response", e))
        }
    }

    /**
     * Convert generic exceptions to domain-specific exceptions
     */
    private fun handleException(exception: Throwable): RebrickableApiException {
        return when (exception) {
            is RebrickableApiException -> exception
            else -> NetworkException("Network error: ${exception.message}", exception)
        }
    }

    /**
     * Cleanup resources when client is no longer needed
     */
    fun close() {
        httpClient.close()
    }
}